stan/math/prim/scal/fun/grad_2F1.hpp

```diff
@@ -1,42 +1,71 @@
 #ifndef STAN_MATH_PRIM_SCAL_FUN_GRAD_2F1_HPP
 #define STAN_MATH_PRIM_SCAL_FUN_GRAD_2F1_HPP
 
+#include <stan/math/prim/scal/err/domain_error.hpp>
+#include <stan/math/prim/scal/fun/is_nan.hpp>
 #include <cmath>
 
 namespace stan {
   namespace math {
 
-    // Gradient of the hypergeometric function 2F1(a, b | c | z)
-    // with respect to a and c
+    /**
+     * Gradient of the hypergeometric function, 2F1(a1, a2, b1, z)
+     * with respect to a1 and b1 only.
+     *
+     * The generalized hypergeometric function is a power series. This
+     * implementation computes gradient by computing the power series
+     * directly stopping when the series converges to within
+     * <code>precision</code> or takes <code>max_steps</code>.
+     *
+     * If more than <code>max_steps</code> are taken without
+     * converging, the function will throw a domain_error.
+     *
+     * @tparam T type of arguments and result
+     * @param[out] gradA1 output argument for partial w.r.t. a1
+     * @param[out] gradB1 output argument for partial w.r.t. b1
+     * @param[in] a1 a1, see generalized hypergeometric function definition.
+     * @param[in] a2 a2, see generalized hypergeometric function definition.
+     * @param[in] b1 b1, see generalized hypergeometric function definition.
+     * @param[in] z z, see generalized hypergeometric function definition.
+     * @param[in] precision precision of the infinite sum. defaults to 1e-6
+     * @param[in] max_steps number of steps to take. defaults to 10000
+     * @throw throws std::domain_error if not converged after max_steps
+     *
+     */
     template<typename T>
-    void grad_2F1(T& gradA, T& gradC, T a, T b, T c, T z, T precision = 1e-6) {
+    void grad_2F1(T& gradA1, T& gradB1, const T& a1, const T& a2,
+      const T& b1, const T& z, T precision = 1e-6, int max_steps = 1e5) {
       using std::fabs;
 
-      gradA = 0;
-      gradC = 0;
+      gradA1 = 0;
+      gradB1 = 0;
 
-      T gradAold = 0;
-      T gradCold = 0;
+      T gradA1old = 0;
+      T gradB1old = 0;
 
       int k = 0;
-      T tDak = 1.0 / (a - 1);
+      T tDak = 1.0 / (a1 - 1);
 
-      while (fabs(tDak * (a + (k - 1)) ) > precision || k == 0) {
-          const T r = ( (a + k) / (c + k) ) * ( (b + k) / (T)(k + 1) ) * z;
-          tDak = r * tDak * (a + (k - 1)) / (a + k);
+      do {
```

Comment:
Seems really clear for algorithms that hit the loop at least once but I'm fine changing it.